Kernel/Net/TCPSocket.cpp

```diff
@@ -414,19 +415,45 @@ NetworkOrdered<u16> TCPSocket::compute_tcp_checksum(IPv4Address const& source, I
 
 ErrorOr<void> TCPSocket::protocol_bind()
 {
-    return m_adapter.with([this](auto& adapter) -> ErrorOr<void> {
+    dbgln_if(TCP_SOCKET_DEBUG, "TCPSocket::protocol_bind(), local_port() is {}", local_port());
+    // Check that we do have the address we're trying to bind to.
+    TRY(m_adapter.with([this](auto& adapter) -> ErrorOr<void> {
         if (has_specific_local_address() && !adapter) {
             adapter = NetworkingManagement::the().from_ipv4_address(local_address());
             if (!adapter)
                 return set_so_error(EADDRNOTAVAIL);
         }
         return {};
-    });
-}
-
-ErrorOr<void> TCPSocket::protocol_listen(bool did_allocate_port)
-{
-    if (!did_allocate_port) {
+    }));
+
+    if (local_port() == 0) {
+        // Allocate an unused ephemeral port.
+        constexpr u16 first_ephemeral_port = 32768;
+        constexpr u16 last_ephemeral_port = 60999;
+        constexpr u16 ephemeral_port_range_size = last_ephemeral_port - first_ephemeral_port;
+        u16 first_scan_port = first_ephemeral_port + get_good_random<u16>() % ephemeral_port_range_size;
```

Comment:
Nit: The randomness will not be uniform here since 65536 mod (60999 - 32768) = 9074 but should be 0 for uniformity. Overall I believe the first ~9000 ports or so will be a little more likely to be picked than the others out of ~28000.